### PR TITLE
[FLINK-29922] Support create external table for hive catalog

### PR DESCRIPTION
**Pull request by wxplovecc**
support hive catalog to create external table

### CODE REVIEW

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Thanks wxplovecc for the contribution.

Can you add a test case for `create external table`?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/CatalogOptions.java [hunk 1]
```diff
@@ -49,6 +49,12 @@ private CatalogOptions() {}
                     .noDefaultValue()
                     .withDescription("Uri of metastore server.");
 
+    public static final ConfigOption<Boolean> HMS_EXTERNAL =
+            ConfigOptions.key("hms.external")
```
Comment (SteNicholas) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/CatalogOptions.java:53]: Could this introduce the `TableType` enum and `table.type` config option which default value is `MANAGED_TABLE`? cc @JingsongLi

Comment (wxplovecc) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/CatalogOptions.java]: @SteNicholas Thanks for review, updated

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/CatalogOptions.java [hunk 2]
```diff
@@ -49,6 +50,12 @@ private CatalogOptions() {}
                     .noDefaultValue()
                     .withDescription("Uri of metastore server.");
 
+    public static final ConfigOption<TableType> TABLE_TYPE =
+            ConfigOptions.key("table.type")
+                    .enumType(TableType.class)
+                    .defaultValue(TableType.MANAGED_TABLE)
+                    .withDescription("Table type of hms table, MANAGED_TABLE/EXTERNAL_TABLE.");
```
Comment (SteNicholas) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/CatalogOptions.java:57]:
```suggestion
                    .withDescription("Type of table.");
```

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.store.table;
+
+/** TableType enum of catalog table. */
```
Comment (SteNicholas) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java:21]:
```suggestion
/** Enum of catalog table type. */
```

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.store.table;
+
+/** Enum of catalog table type. */
+public enum TableType {
+    MANAGED_TABLE,
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java:23]: `MANAGED` is OK.
Can you let it implements `DescribedEnum` just like `ChangelogProducer`?

File: flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java [hunk 5]
```diff
@@ -226,6 +227,13 @@ public void createTable(ObjectPath tablePath, UpdateSchema updateSchema, boolean
                     e);
         }
         Table table = newHmsTable(tablePath);
+
+        if (hiveConf.getEnum(TABLE_TYPE.key(), TableType.MANAGED_TABLE)
+                .equals(TableType.EXTERNAL_TABLE)) {
+            table.setTableType(TableType.EXTERNAL_TABLE.toString());
```
Comment (SteNicholas) [inline flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java:233]:
```suggestion
            table.setTableType(TableType.EXTERNAL_TABLE.name());
```

File: flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java [hunk 6]
```diff
@@ -226,6 +227,13 @@ public void createTable(ObjectPath tablePath, UpdateSchema updateSchema, boolean
                     e);
         }
         Table table = newHmsTable(tablePath);
+
+        if (hiveConf.getEnum(TABLE_TYPE.key(), TableType.MANAGED_TABLE)
```
Comment (SteNicholas) [inline flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java:231]: It's better to move this logic to `newHmsTable` which initialzes the `Table` and sets the parameters of the `Table`.

Comment (wxplovecc) [reply flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java]: done

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java (hunk 4 above)
Comment (wxplovecc) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java]: updated

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+package org.apache.flink.table.store.table;
+
+import org.apache.flink.configuration.DescribedEnum;
+import org.apache.flink.configuration.description.InlineElement;
+
+import static org.apache.flink.configuration.description.TextElement.text;
+
+/** Enum of catalog table type. */
+public enum TableType implements DescribedEnum {
+    MANAGED("MANAGED_TABLE", "Hive manage the lifecycle of the table."),
```
Comment (SteNicholas) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java:28]:
```suggestion
    MANAGED("MANAGED_TABLE", "Table Store owns the table where the entire lifecycle of the table data is managed."),
```

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    EXTERNAL("EXTERNAL_TABLE", "Files are already present or in remote locations.");
```
Comment (SteNicholas) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java:29]:
```suggestion
    EXTERNAL("EXTERNAL_TABLE", "The table where Table Store has loose coupling with the data stored in external locations.");
```

Comment (wxplovecc) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/table/TableType.java]: updated @SteNicholas

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Review by JingsongLi — APPROVED** [review]
Looks good to me!

**Pull request merged**